### PR TITLE
(GH-2629) Add top-level metaparameter keys to YAML plan steps

### PR DESCRIPTION
**Pull request by beechtom**
This adds keys to most YAML plan steps that correspond to plan function
metaparameters. For example, the script step now accepts an `env_vars`
key that corresponds to the `_env_vars` metaparameter, and allows users
to set environment variables on a target when executing the script.

This also fixes a bug where targets listed under the `targets` key in a
plan step were not being passed to the plan. Targets specified in this
manner are now passed to the plan correctly.

Lastly, this includes a refactoring of the `Bolt::PAL::YamlPlan::Step`
classes and the `Bolt::PAL::YamlPlan::Evaluator` class. The step classes
now include methods for formatting options and arguments for plan
function calls as well as an `#evaluate` method that handles evaluating
the step body and calling the step's plan function. The evaluator now
calls out to each step's `#evaluate` function to evaluate the step
instead of including separate logic for evaluating each step and
formatting the plan function call. The motivation for this refactoring
was to keep all of the logic for a given step in the step class, as
opposed to spreading it across multiple classes with duplicate logic.

!feature

* **Support metaparameters as top-level keys in YAML plan steps**
([#2629](https://github.com/puppetlabs/bolt/issues/2629))

YAML plan steps now support metaparameters as top-level keys. For
example, the `script` step supports an `env_vars` key which accepts a
hash of environment variables to set on the target when running the
script.

!bug

* **Run YAML plan `plan` steps with `targets` key**

YAML plans that have a `plan` step with a top-level `targets` key now
pass the targets to the plan.

### CODE REVIEW

File: documentation/writing_yaml_plans.md [hunk 1]
```diff
@@ -77,40 +77,37 @@ Each plan name segment must begin with a lowercase letter and:
...
-failed.
-
-Steps use these fields:
-
--   `name`: (Optional) A unique name that can be used to refer to the result of
-    the step later.
--   `description`: (Optional) An explanation of what the step is doing.
+When the plan runs, each step is executed in order. By default, if a step fails
+the plan halts execution and raises an error containing the result of the step
+that failed.
```
Comment (hestonhoffman) [inline documentation/writing_yaml_plans.md:98]:
```suggestion
When the plan runs, each step is executed in order. If a step fails,
the plan halts execution and raises an error containing the result of the step
that failed.
```
Do we need the "by default" here? It's a bit awkward with the conditional cos you need another comma.

File: documentation/writing_yaml_plans.md [hunk 2]
```diff
@@ -77,40 +77,37 @@ Each plan name segment must begin with a lowercase letter and:
...
 
-Other available keys depend on the type of step.
-
-#### Message step
+### Message step
 
 Use a `message` step to print a message. This will print a message to standard
 out (stdout) when using the `human` output format, and print to standard error
 (stderr) when using the `json` output format.
```
Comment (hestonhoffman) [inline documentation/writing_yaml_plans.md:104]:
```suggestion
Use a `message` step to print a message. The step prints a message to standard
out (stdout) when using the `human` output format, and prints to standard error
(stderr) when using the `json` output format. 
```

File: documentation/writing_yaml_plans.md [hunk 3]
```diff
@@ -127,17 +124,23 @@ prints the object as a string.
...
+code of any command is non-zero.
 
-Command steps use these fields:
+Command steps support the following keys:
 
--   `command`: The command to run
--   `targets`: A target or list of targets to run the command on
+| Key | Type | Description | Required |
+| --- | --- | --- | --- |
+| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |
```
Comment (hestonhoffman) [inline documentation/writing_yaml_plans.md:137]:
```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

File: documentation/writing_yaml_plans.md [hunk 4]
```diff
@@ -151,15 +154,22 @@ steps:
...
 
-Task steps use these fields:
+Task steps support the following keys:
 
--   `task`: The task to run
--   `targets`: A target or list of targets to run the task on
--   `parameters`: (Optional) A map of parameter values to pass to the task
+| Key | Type | Description | Required |
+| --- | --- | --- | --- |
+| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |
```
Comment (hestonhoffman) [inline documentation/writing_yaml_plans.md:165]:
```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

File: documentation/writing_yaml_plans.md [hunk 5]
```diff
@@ -276,14 +308,21 @@ steps:
...
 Use a `plan` step to run another plan and save its result.
 
-Plan steps use these fields:
+Plan steps support the following keys:
 
--   `plan`: The name of the plan to run
--   `parameters`: (Optional) A map of parameter values to pass to the plan
+| Key | Type | Description | Required |
+| --- | --- | --- | --- |
+| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |
```
Comment (hestonhoffman) [inline documentation/writing_yaml_plans.md:319]:
```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

File: documentation/writing_yaml_plans.md [hunk 6]
```diff
@@ -310,10 +349,17 @@ the targets specified with the `targets` field. For more information about
...
 section.
 
-Resources steps use these fields:
+Resources steps support the following keys:
 
--   `resources`: An array of resources to apply
--   `targets`: A target or list of targets to apply the resources on
+| Key | Type | Description | Required |
+| --- | --- | --- | --- |
+| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |
```
Comment (hestonhoffman) [inline documentation/writing_yaml_plans.md:356]:
```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

**Review by hestonhoffman — COMMENTED** [review]
Looking great. I think the tables are a really good addition

File: documentation/writing_yaml_plans.md [hunk 7]
```diff
@@ -77,40 +77,37 @@ Each plan name segment must begin with a lowercase letter and:
 
 ## Plan structure
 
-YAML plans contain a list of steps with optional parameters and results.
+YAML plans contain a list of steps and can include additional optional keys.
```
Comment (lucywyman) [inline documentation/writing_yaml_plans.md:80]:
```suggestion
YAML plans contain a list of steps and can include optional top-level keys.
```

File: documentation/writing_yaml_plans.md [hunk 8]
```diff
@@ -77,40 +77,37 @@ Each plan name segment must begin with a lowercase letter and:
...
+The following top-level keys are available:
```
Comment (lucywyman) [inline documentation/writing_yaml_plans.md:81]: These formatting improvements are SO MUCH BETTER! I feel like at some point we'll want to add a reference doc for YAML plan keys, but definitely doesn't have to be here (I'm also like 99% sure you've had the same idea :stuck_out_tongue_winking_eye:)

Comment (beechtom) [reply documentation/writing_yaml_plans.md:81]: You know me so well 😅

File: lib/bolt/pal/yaml_plan/step.rb [hunk 9]
```diff
@@ -24,31 +20,92 @@ def self.allowed_keys
...
+        #
+        def evaluate(scope, evaluator)
+          evaluated = evaluator.evaluate_code_blocks(scope, body)
+          scope.call_function(function, format_args(evaluated))
+        end
+
+        # Formats a list of args from the provided body
+        #
+        private def format_args(_body)
+          raise NotImplementedError, "Step class #{self.class} does not implement #args"
```
Comment (lucywyman) [inline lib/bolt/pal/yaml_plan/step.rb:89]:
```suggestion
          raise NotImplementedError, "Step class #{self.class} does not implement #format_args"
```

File: lib/bolt/pal/yaml_plan/step/script.rb [hunk 10]
```diff
@@ -6,35 +6,34 @@ class YamlPlan
...
         class Script < Step
           def self.allowed_keys
-            super + Set['script', 'parameters', 'arguments']
+            super + Set['arguments']
           end
 
-          def self.required_keys
-            Set['targets']
+          def self.option_keys
+            Set['catch_errors', 'env_vars', 'run_as']
```
Comment (lucywyman) [inline lib/bolt/pal/yaml_plan/step/script.rb:13]: This is an issue for Puppet plans too, but we don't validate that `env_vars` is a hash before trying to operate on it. Totally up to you if you want to ticket it, or solve it here.
Plan:
```
  - name: script_step
    script: /home/lucy/githubs/bolt/tasks/init.sh
    env_vars: ''
    targets: $targets
```
Output:
```
...
Failed on localhost:20024:
  undefined method `map' for "":String
  Did you mean?  tap
Failed on localhost:20025:
  undefined method `map' for "":String
  Did you mean?  tap
Failed on 2 targets: localhost:20024,localhost:20025
...
```

File: lib/bolt/pal/yaml_plan/step/eval.rb [hunk 11]
```diff
@@ -5,28 +5,28 @@ class PAL
...
+          # Evaluates the step
+          #
+          def evaluate(scope, evaluator)
+            evaluated = evaluator.evaluate_code_blocks(scope, body)
+            evaluated['eval']
           end
 
+          # Transpiles the step into the plan language
+          #
           def transpile
```
Comment (lucywyman) [inline lib/bolt/pal/yaml_plan/step/eval.rb:21]: This is a funny case, but if this is `null` it's valid in the YAML plan but stacktraces when transpiled:
```
...
steps:
  - eval: null
...
```
...
/home/lucy/githubs/bolt/lib/bolt/pal/yaml_plan/step/eval.rb:33:in `transpile': no implicit conversion of nil into String (TypeError)
```

**Review by lucywyman — APPROVED** [review]
### Test cases
- [x] Passes `targets` key in plan step to the plan
- [x] Errors with metaparameter keys misspelled
- [x] Errors when users specify metaparameters under the `parameters` task step key 
- Correctly uses metaparameters in YAML steps
  - Command
    - [x] _catch_errors
    - [x] _run_as
    - [x] _env_vars
  - Task
    - [x] _catch_errors
    - [x] _run_as
    - [x] _noop
  - Script
    - [x] _catch_errors
    - [x] _run_as
    - [x] _env_vars
  - Download
    - [x] _catch_errors
    - [] _run_as
  - Upload
    - [x] _catch_errors
    - [x] _run_as
  - Plan
    - [x] _catch_errors
    - [x] _run_as
  - Resources
    - [x] _catch_errors
    - [x] _run_as
    - [x] _noop
    - [x] _description
- [x] Correctly transpiles metaparameters to Puppet plans
- [x] Correctly transpiles functions without metaparameters to Puppet plans
- [x] Correctly transpiles functions with descriptions
- [x] Merges options with plan and task parameters when transpiling

Looks great!

File: lib/bolt/pal/yaml_plan/step/script.rb (hunk 10 above)
Comment (beechtom) [reply lib/bolt/pal/yaml_plan/step/script.rb:13]: I'll ticket this separately.

File: lib/bolt/pal/yaml_plan/step/eval.rb (hunk 11 above)
Comment (beechtom) [reply lib/bolt/pal/yaml_plan/step/eval.rb:21]: Fixed.

**Pull request merged**